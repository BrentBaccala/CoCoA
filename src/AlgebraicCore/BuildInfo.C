//   Copyright (c)  2005,2007,2012  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/config.H"

#include <limits>
#include <iostream>
using std::ostream;
using std::endl;
#include <string>
using std::string;

namespace CoCoA
{

  namespace BuildInfo
  {
    // Build info is actually passed in via three preprocessor variables:
    // COCOA_VERSION, COCOA_CXX, COCOA_CXXFLAGS -- all three are quoted strings.

    const std::string& version()
    {
      static const string info(COCOA_VERSION);
      return info;
    }

    const std::string& compiler()
    {
      static const string info(COCOA_CXX);
      return info;
    }

    const std::string& CompilationFlags()
    {
      static const string info(COCOA_CXXFLAGS);
      return info;
    }

    const std::string& CompilationDefines()
    {
      static const string info(COCOA_CXXFLAGS_DEFINES);
      return info;
    }


    void PrintAll(std::ostream& out)
    {
    // Below we use string literal juxtaposition, so that the build
    // info can readily be extracted from libcocoa.a by the command
    // strings libcocoa.a | egrep "CoCoA::BuildInfo"
      out << endl
          << "CoCoA::BuildInfo Summary of build information for CoCoALib:" << endl
          << "CoCoA::BuildInfo CoCoALib Version: " COCOA_VERSION << endl
          << "CoCoA::BuildInfo Compiler: " COCOA_CXX << endl
          << "CoCoA::BuildInfo Compilation Flags: " COCOA_CXXFLAGS << endl
          << "CoCoA::BuildInfo Compilation Flags Defines: " COCOA_CXXFLAGS_DEFINES << endl
          << "CoCoA::BuildInfo NumBits int = " << std::numeric_limits<unsigned int>::digits << endl
          << "CoCoA::BuildInfo NumBits long = " << std::numeric_limits<unsigned long>::digits << endl
          << "CoCoA::BuildInfo NumBits SmallExponent_t = " << std::numeric_limits<SmallExponent_t>::digits << endl
          << endl;
    }

  } // end of namespace CoCoA::BuildInfo

} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/BuildInfo.C,v 1.8 2012/10/24 12:10:45 abbott Exp $
// $Log: BuildInfo.C,v $
// Revision 1.8  2012/10/24 12:10:45  abbott
// Revised PrintAll: now prints also NumBits of SmallExponent_t
//
// Revision 1.7  2012/10/02 11:45:09  bigatti
// -- added CompilationDefines
//
// Revision 1.6  2012/10/02 10:36:12  abbott
// Revised interface to BuildInfo information strings.
// Several consequential changes.
//
// Revision 1.5  2012/06/20 12:26:26  bigatti
// -- now exporting also compiler and compilation flags
//
// Revision 1.4  2011/03/11 17:38:21  bigatti
// -- added missing include
//
// Revision 1.3  2011/03/11 17:18:55  bigatti
// -- added info on int and long size
//
// Revision 1.2  2007/10/30 17:14:08  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.4  2007/02/10 18:44:03  cocoa
// Added "const" twice to each test and example.
// Eliminated dependency on io.H in several files.
// Improved BuildInfo, and added an example about how to use it.
// Some other minor cleaning.
//
// Revision 1.3  2007/02/08 22:34:22  cocoa
// Changed BuildInfo: only BuildInfo version is publicly visible.
// Only BuildInfo needs complicated compile-time flags, so several changes
// to Makefiles etc.  Added a new example: ex-BuildInfo.C
//
// Revision 1.1.1.1  2005/10/17 10:46:54  cocoa
// Imported files
//
// Revision 1.2  2005/10/12 15:52:09  cocoa
// Completed test-RingFp1 and corrected/cleaned the SmallFp*
// and RingFp* files.
//
// Some minor tidying elsewhere.
//
// Revision 1.1  2005/10/06 16:36:42  cocoa
// Added the capability find out build information at run-time.
// The Makefiles should be a little tidier too.
//
