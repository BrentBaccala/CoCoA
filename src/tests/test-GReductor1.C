//   Copyright (c)  2007  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/GlobalManager.H"
#include "CoCoA/BuildInfo.H"
#include "CoCoA/FractionField.H"
#include "CoCoA/FreeModule.H"
#include "CoCoA/IntOperations.H"
#include "CoCoA/TmpGOperations.H"
#include "CoCoA/TmpGPoly.H"
#include "CoCoA/ModuleOrdering.H"
#include "CoCoA/QuotientRing.H"
#include "CoCoA/RingDistrMPolyClean.H"
#include "CoCoA/RingDistrMPolyInlFpPP.H"
#include "CoCoA/RingDistrMPolyInlPP.H"
#include "CoCoA/RingFp.H"
#include "CoCoA/RingQQ.H"
#include "CoCoA/SparsePolyRing.H"
#include "CoCoA/VectorOperations.H"
#include "CoCoA/submodule.H"
#include "CoCoA/symbol.H"

using namespace CoCoA;

#include <algorithm>
using std::find_if;
#include <functional>
using std::not1;
using std::ptr_fun;
#include <iostream>
using std::cout;
using std::cerr;
using std::endl;
using std::boolalpha;
#include <string>
#include <vector>
using std::vector;

//----------------------------------------------------------------------
// Test for Buchberger's Algorithm 
// functions: ComputeGBasis
// environments: DMP (Fp, Q), DMPI (Fp), DMPII
//----------------------------------------------------------------------

#define TEST_ASSERT(cond)   do { if (!(cond)) CoCoA::AssertionFailed(#cond, __FILE__, __LINE__); } while(0)

typedef bool (*bool_RE)(ConstRefRingElem);
bool HasHomogElems(const vector<RingElem>& l)
// We *DO NOT USE* STL algorithm because std::ptr_fun fails when fn has formal params of reference type
//{ return find_if(++l.begin(), l.end(), not1(ptr_fun(static_cast<bool_RE>(IsHomog)))) == l.end(); }
{
  const long n = len(l);
  for (long i=0; i < n; ++i)
    if (!IsHomog(l[i])) return false;
  return true;
}


void PrintInfo(const std::string& s, const SparsePolyRing& P, const vector<RingElem>& InputPolys)
{
  cout << "Executing test " << s << endl << P << "   ";
  cout << "GradingDim = " << GradingDim(P) << "   ";
  if (GradingDim(P)!=0)
    cout << "IsHomog(InputPolys) = " << HasHomogElems(InputPolys);
  cout << endl;
}


// expecting P to have 5 indeterminates
void testC4_h(SparsePolyRing P)
{
  RingElem x = indet(P,0),  y = indet(P,1),  z = indet(P,2), t=indet(P,3);
  RingElem h = indet(P,4);
  PolyList InputPolys,GB;
  InputPolys.push_back(x+y+z+t);
  InputPolys.push_back(x*y+y*z+z*t+t*x);
  InputPolys.push_back(x*y*z+y*z*t+z*t*x+t*x*y);
  InputPolys.push_back(x*y*z*t-power(h,4));
  PrintInfo("C4_h", P, InputPolys);
  PolyList MinGens;
  ComputeGBasis(GB, MinGens, InputPolys);
  TEST_ASSERT(len(MinGens)==4);
  monic(GB);
  cout << "C4_h = " << GB << "\n"<<endl;
}


void testC4(SparsePolyRing P)
{
  RingElem x = indet(P,0),  y = indet(P,1),  z = indet(P,2), t = indet(P,3);
  PolyList InputPolys,GB;
  InputPolys.push_back(x+y+z+t);
  InputPolys.push_back(x*y + y*z + z*t + t*x);
  InputPolys.push_back(x*y*z + y*z*t + z*t*x + t*x*y);
  InputPolys.push_back(x*y*z*t - 1);
  PrintInfo("C4", P, InputPolys);
  PolyList MinGens;
  ComputeGBasis(GB, MinGens, InputPolys);
  TEST_ASSERT(len(MinGens)==0);
  monic(GB);
  cout << "C4 = " << GB << "\n"<<endl;
}

// expecting P to have 4 indeterminates
void testCurve_h(SparsePolyRing P)
{
  RingElem t = indet(P,0),  x = indet(P,1),  y = indet(P,2), z=indet(P,3);
  PolyList InputPolys,GB;
  InputPolys.push_back(power(t,3)-power(x,3));
  InputPolys.push_back(power(t,5)-power(y,5));
  InputPolys.push_back(power(t,7)-power(z,7));
  PrintInfo("Curve_h", P, InputPolys);
  PolyList MinGens;
  ComputeGBasis(GB, MinGens, InputPolys);
  TEST_ASSERT(len(MinGens)==3);
  monic(GB);
  cout << "Curve_h = " << GB << "\n"<<endl;
}

// expecting P to have 5 indeterminates
void testCurve(SparsePolyRing P)
{
  const vector<RingElem>& x = indets(P);
  PolyList InputPolys, GB;
  InputPolys.push_back(power(x[0],3) - x[1]);
  InputPolys.push_back(power(x[0],5) - x[2]);
  InputPolys.push_back(power(x[0],7) - x[3]);
  PrintInfo("Curve", P, InputPolys);
  PolyList MinGens;
  ComputeGBasis(GB, MinGens, InputPolys);
  TEST_ASSERT(len(MinGens)==0);
  monic(GB);
  cout << "Curve = " << GB << "\n" << endl;
}

void program()
{
  GlobalManager CoCoAFoundations(UseNonNegResidues);
  cout << boolalpha;
  
  // This is a test of the correctness of GB computations for the three cases:
  // AllGraded, AllAffine and GradedRing, Inhomogenous Input for ideals only.
  // This does not check if the computations have been done in the right way,
  // only if the result is correct.

  // the results have been tested with cocoa 4
  
  ring ZZmod = NewZZmod(101);  
  
  // Cyclic 4
  const vector<symbol> X = SymbolRange("x", 0, 4);
  SparsePolyRing P(NewPolyRing_DMPI(ZZmod, X));
  SparsePolyRing Q(NewPolyRing_DMPII(ZZmod, X));
  SparsePolyRing R(NewPolyRing(RingQQ(), 5));
  SparsePolyRing T(NewPolyRing_DMPI(RingQQ(), X));

  // Graded Ring, Homogeneous Input
  testC4_h(P);
  testC4_h(Q);
  testC4_h(R);
  testC4_h(T);
 
  // Graded Ring, Non-homogeneous Input
  testC4(P);
  testC4(Q);
  testC4(R);
  testC4(T);
  
  SparsePolyRing LexR = NewPolyRing(RingQQ(),X,NewLexOrdering(5));
  // Non-graded Ring, Non-homogeneous Input
  //testC4(LexR,"NewPolyRing Q");
    
  // Curve
  const vector<symbol> Y = SymbolRange("x", 0 ,3);
  SparsePolyRing P1(NewPolyRing_DMPI(ZZmod, Y));
  SparsePolyRing Q1(NewPolyRing_DMPII(ZZmod, Y));
  SparsePolyRing R1(NewPolyRing(RingQQ(), 4));
  SparsePolyRing T1 (NewPolyRing_DMPI(RingQQ(), Y));

  // Graded Ring, Homogeneous Input
  testCurve_h(P1);
  testCurve_h(Q1);
  testCurve_h(R1);
  testCurve_h(T1);
  
  // Graded Ring, Non-homogeneous Input
  testCurve(P1);
  testCurve(Q1);
  testCurve(R1);
  testCurve(T1);
  
  SparsePolyRing LexR1 = NewPolyRing(RingQQ(),Y,NewLexOrdering(4));

  // Non-graded Ring, Non-homogeneous Input
  testCurve(LexR1);
}


// Use main() to handle any uncaught exceptions and warn the user about them.
int main()
{
  try
  {
    program();
    return 0;
  }
  catch (const CoCoA::ErrorInfo& err)
  {
    cerr << "***ERROR***  UNCAUGHT CoCoA Error";
    ANNOUNCE(cerr, err);
  }
  catch (const std::exception& exc)
  {
    cerr << "***ERROR***  UNCAUGHT std::exception: " << exc.what() << endl;
  }
  catch(...)
  {
    cerr << "***ERROR***  UNCAUGHT UNKNOWN EXCEPTION" << endl;
  }

  BuildInfo::PrintAll(cerr);
  return 1;
}
