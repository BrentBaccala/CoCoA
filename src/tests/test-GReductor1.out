Executing test C4_h
RingWithID(3,"RingWithID(2)[x[0],x[1],x[2],x[3],x[4]]")   GradingDim = 1   IsHomog(InputPolys) = true
C4_h = [x[0] +x[1] +x[2] +x[3],  x[1]^2 +2*x[1]*x[3] +x[3]^2,  x[1]*x[2]^2 +x[2]^2*x[3] +100*x[1]*x[3]^2 +100*x[3]^3,  x[1]*x[2]*x[3]^2 +x[2]^2*x[3]^2 +100*x[1]*x[3]^3 +x[2]*x[3]^3 +100*x[3]^4 +100*x[4]^4,  x[2]^3*x[3]^2 +x[2]^2*x[3]^3 +100*x[2]*x[4]^4 +100*x[3]*x[4]^4,  x[1]*x[3]^4 +x[3]^5 +100*x[1]*x[4]^4 +100*x[3]*x[4]^4,  x[2]^2*x[3]^4 +x[1]*x[2]*x[4]^4 +100*x[1]*x[3]*x[4]^4 +x[2]*x[3]*x[4]^4 +99*x[3]^2*x[4]^4]

Executing test C4_h
RingWithID(4,"RingWithID(2)[x[0],x[1],x[2],x[3],x[4]]")   GradingDim = 1   IsHomog(InputPolys) = true
C4_h = [x[0] +x[1] +x[2] +x[3],  x[1]^2 +2*x[1]*x[3] +x[3]^2,  x[1]*x[2]^2 +x[2]^2*x[3] +100*x[1]*x[3]^2 +100*x[3]^3,  x[1]*x[2]*x[3]^2 +x[2]^2*x[3]^2 +100*x[1]*x[3]^3 +x[2]*x[3]^3 +100*x[3]^4 +100*x[4]^4,  x[2]^3*x[3]^2 +x[2]^2*x[3]^3 +100*x[2]*x[4]^4 +100*x[3]*x[4]^4,  x[1]*x[3]^4 +x[3]^5 +100*x[1]*x[4]^4 +100*x[3]*x[4]^4,  x[2]^2*x[3]^4 +x[1]*x[2]*x[4]^4 +100*x[1]*x[3]*x[4]^4 +x[2]*x[3]*x[4]^4 +99*x[3]^2*x[4]^4]

Executing test C4_h
RingWithID(5,"QQ[x[0],x[1],x[2],x[3],x[4]]")   GradingDim = 1   IsHomog(InputPolys) = true
C4_h = [x[0] +x[1] +x[2] +x[3],  x[1]^2 +2*x[1]*x[3] +x[3]^2,  x[1]*x[2]^2 +x[2]^2*x[3] -x[1]*x[3]^2 -x[3]^3,  x[1]*x[2]*x[3]^2 +x[2]^2*x[3]^2 -x[1]*x[3]^3 +x[2]*x[3]^3 -x[3]^4 -x[4]^4,  x[2]^3*x[3]^2 +x[2]^2*x[3]^3 -x[2]*x[4]^4 -x[3]*x[4]^4,  x[1]*x[3]^4 +x[3]^5 -x[1]*x[4]^4 -x[3]*x[4]^4,  x[2]^2*x[3]^4 +x[1]*x[2]*x[4]^4 -x[1]*x[3]*x[4]^4 +x[2]*x[3]*x[4]^4 -2*x[3]^2*x[4]^4]

Executing test C4_h
RingWithID(6,"QQ[x[0],x[1],x[2],x[3],x[4]]")   GradingDim = 1   IsHomog(InputPolys) = true
C4_h = [x[0] +x[1] +x[2] +x[3],  x[1]^2 +2*x[1]*x[3] +x[3]^2,  x[1]*x[2]^2 +x[2]^2*x[3] -x[1]*x[3]^2 -x[3]^3,  x[1]*x[2]*x[3]^2 +x[2]^2*x[3]^2 -x[1]*x[3]^3 +x[2]*x[3]^3 -x[3]^4 -x[4]^4,  x[2]^3*x[3]^2 +x[2]^2*x[3]^3 -x[2]*x[4]^4 -x[3]*x[4]^4,  x[1]*x[3]^4 +x[3]^5 -x[1]*x[4]^4 -x[3]*x[4]^4,  x[2]^2*x[3]^4 +x[1]*x[2]*x[4]^4 -x[1]*x[3]*x[4]^4 +x[2]*x[3]*x[4]^4 -2*x[3]^2*x[4]^4]

Executing test C4
RingWithID(3,"RingWithID(2)[x[0],x[1],x[2],x[3],x[4]]")   GradingDim = 1   IsHomog(InputPolys) = false
C4 = [x[0] +x[1] +x[2] +x[3],  x[1]^2 +2*x[1]*x[3] +x[3]^2,  x[1]*x[2]^2 +x[2]^2*x[3] +100*x[1]*x[3]^2 +100*x[3]^3,  x[1]*x[2]*x[3]^2 +x[2]^2*x[3]^2 +100*x[1]*x[3]^3 +x[2]*x[3]^3 +100*x[3]^4 +100,  x[2]^3*x[3]^2 +x[2]^2*x[3]^3 +100*x[2] +100*x[3],  x[1]*x[3]^4 +x[3]^5 +100*x[1] +100*x[3],  x[2]^2*x[3]^4 +x[1]*x[2] +100*x[1]*x[3] +x[2]*x[3] +99*x[3]^2]

Executing test C4
RingWithID(4,"RingWithID(2)[x[0],x[1],x[2],x[3],x[4]]")   GradingDim = 1   IsHomog(InputPolys) = false
C4 = [x[0] +x[1] +x[2] +x[3],  x[1]^2 +2*x[1]*x[3] +x[3]^2,  x[1]*x[2]^2 +x[2]^2*x[3] +100*x[1]*x[3]^2 +100*x[3]^3,  x[1]*x[2]*x[3]^2 +x[2]^2*x[3]^2 +100*x[1]*x[3]^3 +x[2]*x[3]^3 +100*x[3]^4 +100,  x[2]^3*x[3]^2 +x[2]^2*x[3]^3 +100*x[2] +100*x[3],  x[1]*x[3]^4 +x[3]^5 +100*x[1] +100*x[3],  x[2]^2*x[3]^4 +x[1]*x[2] +100*x[1]*x[3] +x[2]*x[3] +99*x[3]^2]

Executing test C4
RingWithID(5,"QQ[x[0],x[1],x[2],x[3],x[4]]")   GradingDim = 1   IsHomog(InputPolys) = false
C4 = [x[0] +x[1] +x[2] +x[3],  x[1]^2 +2*x[1]*x[3] +x[3]^2,  x[1]*x[2]^2 +x[2]^2*x[3] -x[1]*x[3]^2 -x[3]^3,  x[1]*x[2]*x[3]^2 +x[2]^2*x[3]^2 -x[1]*x[3]^3 +x[2]*x[3]^3 -x[3]^4 -1,  x[2]^3*x[3]^2 +x[2]^2*x[3]^3 -x[2] -x[3],  x[1]*x[3]^4 +x[3]^5 -x[1] -x[3],  x[2]^2*x[3]^4 +x[1]*x[2] -x[1]*x[3] +x[2]*x[3] -2*x[3]^2]

Executing test C4
RingWithID(6,"QQ[x[0],x[1],x[2],x[3],x[4]]")   GradingDim = 1   IsHomog(InputPolys) = false
C4 = [x[0] +x[1] +x[2] +x[3],  x[1]^2 +2*x[1]*x[3] +x[3]^2,  x[1]*x[2]^2 +x[2]^2*x[3] -x[1]*x[3]^2 -x[3]^3,  x[1]*x[2]*x[3]^2 +x[2]^2*x[3]^2 -x[1]*x[3]^3 +x[2]*x[3]^3 -x[3]^4 -1,  x[2]^3*x[3]^2 +x[2]^2*x[3]^3 -x[2] -x[3],  x[1]*x[3]^4 +x[3]^5 -x[1] -x[3],  x[2]^2*x[3]^4 +x[1]*x[2] -x[1]*x[3] +x[2]*x[3] -2*x[3]^2]

Executing test Curve_h
RingWithID(12,"RingWithID(2)[x[0],x[1],x[2],x[3]]")   GradingDim = 1   IsHomog(InputPolys) = true
Curve_h = [x[0]^3 +100*x[1]^3,  x[0]^2*x[1]^3 +100*x[2]^5,  x[1]^6 +100*x[0]*x[2]^5,  x[0]^2*x[2]^5 +100*x[3]^7,  x[1]^3*x[2]^5 +100*x[0]*x[3]^7,  x[2]^10 +100*x[1]^3*x[3]^7]

Executing test Curve_h
RingWithID(13,"RingWithID(2)[x[0],x[1],x[2],x[3]]")   GradingDim = 1   IsHomog(InputPolys) = true
Curve_h = [x[0]^3 +100*x[1]^3,  x[0]^2*x[1]^3 +100*x[2]^5,  x[1]^6 +100*x[0]*x[2]^5,  x[0]^2*x[2]^5 +100*x[3]^7,  x[1]^3*x[2]^5 +100*x[0]*x[3]^7,  x[2]^10 +100*x[1]^3*x[3]^7]

Executing test Curve_h
RingWithID(14,"QQ[x[0],x[1],x[2],x[3]]")   GradingDim = 1   IsHomog(InputPolys) = true
Curve_h = [x[0]^3 -x[1]^3,  x[0]^2*x[1]^3 -x[2]^5,  x[1]^6 -x[0]*x[2]^5,  x[0]^2*x[2]^5 -x[3]^7,  x[1]^3*x[2]^5 -x[0]*x[3]^7,  x[2]^10 -x[1]^3*x[3]^7]

Executing test Curve_h
RingWithID(15,"QQ[x[0],x[1],x[2],x[3]]")   GradingDim = 1   IsHomog(InputPolys) = true
Curve_h = [x[0]^3 -x[1]^3,  x[0]^2*x[1]^3 -x[2]^5,  x[1]^6 -x[0]*x[2]^5,  x[0]^2*x[2]^5 -x[3]^7,  x[1]^3*x[2]^5 -x[0]*x[3]^7,  x[2]^10 -x[1]^3*x[3]^7]

Executing test Curve
RingWithID(12,"RingWithID(2)[x[0],x[1],x[2],x[3]]")   GradingDim = 1   IsHomog(InputPolys) = false
Curve = [x[0]^3 +100*x[1],  x[0]^2*x[1] +100*x[2],  x[1]^2 +100*x[0]*x[2],  x[0]^2*x[2] +100*x[3],  x[2]^2 +100*x[1]*x[3],  x[1]*x[2] +100*x[0]*x[3]]

Executing test Curve
RingWithID(13,"RingWithID(2)[x[0],x[1],x[2],x[3]]")   GradingDim = 1   IsHomog(InputPolys) = false
Curve = [x[0]^3 +100*x[1],  x[0]^2*x[1] +100*x[2],  x[1]^2 +100*x[0]*x[2],  x[0]^2*x[2] +100*x[3],  x[2]^2 +100*x[1]*x[3],  x[1]*x[2] +100*x[0]*x[3]]

Executing test Curve
RingWithID(14,"QQ[x[0],x[1],x[2],x[3]]")   GradingDim = 1   IsHomog(InputPolys) = false
Curve = [x[0]^3 -x[1],  x[0]^2*x[1] -x[2],  x[1]^2 -x[0]*x[2],  x[0]^2*x[2] -x[3],  x[2]^2 -x[1]*x[3],  x[1]*x[2] -x[0]*x[3]]

Executing test Curve
RingWithID(15,"QQ[x[0],x[1],x[2],x[3]]")   GradingDim = 1   IsHomog(InputPolys) = false
Curve = [x[0]^3 -x[1],  x[0]^2*x[1] -x[2],  x[1]^2 -x[0]*x[2],  x[0]^2*x[2] -x[3],  x[2]^2 -x[1]*x[3],  x[1]*x[2] -x[0]*x[3]]

Executing test Curve
RingWithID(20,"QQ[x[0],x[1],x[2],x[3]]")   GradingDim = 0   
Curve = [x[0]^3 -x[1],  x[0]^2*x[1] -x[2],  x[0]*x[2] -x[1]^2,  x[0]*x[1]^2 -x[3],  x[1]^4 -x[2]*x[3],  x[1]*x[3] -x[2]^2,  x[0]*x[3] -x[1]*x[2],  x[1]^3*x[2] -x[3]^2,  x[1]^2*x[2]^3 -x[3]^3,  x[1]*x[2]^5 -x[3]^4,  x[2]^7 -x[3]^5]

